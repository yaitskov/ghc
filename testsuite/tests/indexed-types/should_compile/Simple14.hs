{-# LANGUAGE TypeFamilies, RankNTypes, FlexibleContexts, ScopedTypeVariables #-}

module Simple14 where

data EQ_ x y = EQ_

-- Nov 2014: actually eqE has an ambiguous type
-- Apr 2020: now it doesn't again
eqE :: EQ_ x y -> (x~y => EQ_ z z) -> p
eqE x y = error "eqE"

eqI :: EQ_ x x
eqI = error "eqI"

ntI :: (forall p. EQ_ x y -> p) -> EQ_ x y
ntI x = error "ntI"

foo :: forall m n. EQ_ (Maybe m) (Maybe n)
foo = ntI (\x -> x `eqE` (eqI :: EQ_ m n))

-- Alternative
-- foo = ntI (\eq -> eq `eqE` (eqI :: EQ_ m n))

-- eq :: EQ_ (Maybe m) (Maybe n)
-- Need (Maybe m ~ Maybe n) =>  EQ_ m n ~ EQ_ zeta zeta
-- which reduces to (m~n) => m ~ zeta
-- but then we are stuck
