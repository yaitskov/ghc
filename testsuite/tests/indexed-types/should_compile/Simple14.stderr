
Simple14.hs:19:27: error:
    • Couldn't match type ‘z0’ with ‘n’
      Expected: EQ_ z0 z0
        Actual: EQ_ m n
        ‘z0’ is untouchable
          inside the constraints: Maybe m ~ Maybe n
          bound by a type expected by the context:
                     (Maybe m ~ Maybe n) => EQ_ z0 z0
          at Simple14.hs:19:26-41
      ‘n’ is a rigid type variable bound by
        the type signature for:
          foo :: forall m n. EQ_ (Maybe m) (Maybe n)
        at Simple14.hs:18:1-42
    • In the second argument of ‘eqE’, namely ‘(eqI :: EQ_ m n)’
      In the expression: x `eqE` (eqI :: EQ_ m n)
      In the first argument of ‘ntI’, namely
        ‘(\ x -> x `eqE` (eqI :: EQ_ m n))’
    • Relevant bindings include
        x :: EQ_ (Maybe m) (Maybe n) (bound at Simple14.hs:19:13)
        foo :: EQ_ (Maybe m) (Maybe n) (bound at Simple14.hs:19:1)
